### PR TITLE
Add support for @grpc/grpc-js

### PR DESCRIPTION
**Pull request by badsyntax**
Refs https://github.com/improbable-eng/ts-protoc-gen/issues/226, https://github.com/improbable-eng/ts-protoc-gen/pull/234, https://github.com/grpc/grpc-node/pull/1380

Add support for generating files that import `@grpc/grpc-js` instead of `grpc`. This does not change the default behaviour of the package, but instead introduces a new param allowing you to opt-in to using `@grpc/grpc-js`.

## Changes

- Add new `mode` param, for example:

```sh
--ts_out=service=grpc-node,mode=grpc-js
```

- Add generated examples
- Update README

## Verification

I've only tested it via the generated examples. It's very similar to `service=grpc-node` other than changing the import paths.

### CODE REVIEW

**Comment by laike9m** [comment]
BTW, I tried installing the `grpc-node-mode` branch using

```
npm install badsyntax/ts-protoc-gen#grpc-node-mode
```

run it, and got this error:

```
internal/modules/cjs/loader.js:969
  throw err;
  ^

Error: Cannot find module '../lib/index'
```

It seems the `lib` directory does not exist in `ts-protoc-gen` directory under `node_modules`.

**Comment by badsyntax** [comment]
The lib directory is only built on `prepublishOnly` (before the package is published to npm), so installing via git won't work, see:

https://github.com/improbable-eng/ts-protoc-gen/blob/3e58c65351d72cd7280c0cb16e8f7e758fb88e79/package.json#L15

**Comment by laike9m** [comment]
Thanks for the quick response. Any suggestions on what I should do to make it work? Or should I stick with `sed` for now like https://github.com/badsyntax/grpc-js-types?

**Comment by avdeev** [comment]
Quick fix:

```sh
sed -i -- 's/import \* as grpc from "grpc";/import \* as grpc from "@grpc\/grpc-js";/g' *
```

**Comment by badsyntax** [comment]
if you want to continue using this package, then you're going to have to stick with sed. but as mentioned here https://github.com/grpc/grpc-node/pull/1380#issuecomment-653388835, I would suggest to try the new typescript generator offered by the author of grpc-js

**Comment by AndreyUtka** [comment]
Hi @badsyntax, Any plans to merge this PR?

**Comment by badsyntax** [comment]
I can't merge as I'm not the owner of this repo. Not sure why I haven't got a review yet :(

**Comment by AndreyUtka** [comment]
@MarcusLongmuir Can you please help?

File: src/service/grpcnode.ts
Comment (MarcusLongmuir) [inline src/service/grpcnode.ts:7]: This function is only called from the `index.ts` file so the argument could either be an enum to make the mode explicit.

File: src/index.ts
Comment (MarcusLongmuir) [inline src/index.ts:34]: The `service` parameter should probably have more validation and helpful warnings, but this additional parameter makes that even clearer.

Can you add checks for the supported `service` and `mode` values and change this to throw exceptions if they're unrecognised to help users more quickly identify misconfigurations please?

**Review by MarcusLongmuir — CHANGES_REQUESTED** [review]
Thanks for the contribution! Apologies for the long delay in waiting for a review. This had gone under the radar.

Only a few minor things 👍

**Comment by badsyntax** [comment]
I've made those changes:


Running:

`protoc --plugin="protoc-gen-ts=${PROTOC_GEN_TS_PATH}" --plugin=protoc-gen-grpc=${PROTOC_GEN_GRPC_PATH} --js_out="import_style=commonjs,binary:${OUT_DIR}" --ts_out="service=grpc-nodde:${OUT_DIR}" --grpc_out="${OUT_DIR}" proto/examplecom/simple.proto`

Shows:

```sh
protoc-gen-ts error: Error: Unrecognised service parameter: grpc-nodde
    at getServiceParameter (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:155:19)
    at Object.getParameterEnums (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:173:18)
    at /Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/index.js:19:25
    at Socket.<anonymous> (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:74:9)
    at Socket.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1183:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)

--ts_out: protoc-gen-ts: Plugin failed with status code 1.
```

Running:

`protoc --plugin="protoc-gen-ts=${PROTOC_GEN_TS_PATH}" --plugin=protoc-gen-grpc=${PROTOC_GEN_GRPC_PATH} --js_out="import_style=commonjs,binary:${OUT_DIR}" --ts_out="service=grpc-node,mode=grpc-jss:${OUT_DIR}" --grpc_out="${OUT_DIR}" proto/examplecom/simple.proto`

Shows

```sh
protoc --plugin="protoc-gen-ts=${PROTOC_GEN_TS_PATH}" --plugin=protoc-gen-grpc=${PROTOC_GEN_GRPC_PATH} --js_out="import_style=commonjs,binary:${OUT_DIR}" --ts_out="service=grpc-node,mode=grpc-jss:${OUT_DIR}" --grpc_out="${OUT_DIR}" proto/examplecom/simple.proto
protoc-gen-ts error: Error: Unrecognised mode parameter: grpc-jss
    at getModeParameter (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:166:19)
    at Object.getParameterEnums (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:174:15)
    at /Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/index.js:19:25
    at Socket.<anonymous> (/Users/richardwillis/Projects/badsyntax/ts-protoc-gen/lib/util.js:74:9)
    at Socket.emit (events.js:215:7)
    at endReadableNT (_stream_readable.js:1183:12)
    at processTicksAndRejections (internal/process/task_queues.js:80:21)

--ts_out: protoc-gen-ts: Plugin failed with status code 1.
```

**Comment by badsyntax** [comment]
Awesome, thanks!

**Comment by avdeev** [comment]
@badsyntax Thanks! Could you release it?

**Comment by AndreyUtka** [comment]
thanks !

**Comment by MarcusLongmuir** [comment]
Unfortunately the repo and org setup appears to make that quite difficult without another maintainer so I'll have to do it in the morning when others come online.

If you want to validate usage in the meantime there is now a published version on npm: https://www.npmjs.com/package/ts-protoc-gen/v/0.12.1-pre.6383416.

I'll comment on this PR when the release goes out.

**Comment by MarcusLongmuir** [comment]
`0.13.0` is now out: https://www.npmjs.com/package/ts-protoc-gen/v/0.13.0

**Comment by badsyntax** [comment]
Thanks for the release. I've tested the change [in my project](https://github.com/badsyntax/vscode-gradle/pull/690) and it lgtm 👍

**Pull request merged**
